orm: deterministic versionedIDRef serialization

ruseinov opened the pull request:
Fixes #918 

You will notice that prefix queries should be done by id alone, because there is no way we can consistently deserialise if we ignore `Version` when it is zero.

This also needs bnscli changes, but first I'd like to validate the approach.

iov-bot-danger commented:
<!--
  1 Error: Please include a CHANGELOG ent...
  0 Warnings
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">Please include a CHANGELOG entry. <br />
You can find it at <a href="https://github.com/iov-one/weave/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

ruseinov commented:
@webmaster128 FYI.
The current implementation assumes that a version is a big-endian encoded uint32 into last 4 bytes. The ID does not have fixed length in our spec, so we always need to start with cutting off the last 4 bytes first.
@alpe FYI: The acceptance criteria stated that we need 8 bytes, but since we are using uint32 - I'm not sure why. It's probably smarter to change the type from the get go if we are planning to do this later.

husio commented on orm/versioning_bucket.go:
```diff
@@ -233,3 +226,40 @@ func (marker) Copy() CloneableData {
+
+	res = append(res, key.ID...)
+
+	if omitEmptyVersion && key.Version == 0 {
```
This API can be simplified. ~Versioning starts with 1. This means that `omitEmptyVersion` is not needed. You can use `version == 0` to return key without version. In all cases where `version > 0` returned key should contain the version information.~
If you need only the ID, you can use `VersionedIDRef.ID`. Otherwise you would use `MarshalVersionedID` to serialize the whole reference, together with the version (regardless the value). `omitEmptyVersion` is not needed, because both functionalities can be provided without it.

In the current version `MarshalVersionedID` is not compatible with `UnmarshalVersionedID`.

ruseinov replied:
This was done on purpose, if you look at the commit history I did start with using just the ID, but the I have figured out that for all the external dependencies it is nicer to use this method than doing
```
if id.Version == 0 { .... }
```

It is easier to have a flag in this method than having several places in the code that are using `ID` directly for prefix searches.

While I'm not a big fan of this change I think it is cleaner and more straightforward to use for anyone who might want to utilise these two methods, as both of them are well documented. 

The alternative was to introduce a 3rd method, but that's one too many.

alpe replied:
I am with Piotr on this. `MarshalVersionedID/UnmarshalVersionedID` outputs should be usable to feed each other. This additional `omitEmptyVersion` does not make it more readable to me.
It is only set `true` in the bnscli client and a `gov/bucket_test`
The test could use any random []byte slice to achieve the same results as it is "By unknown key". Just move the related code to bnscli and all good.

husio commented on orm/versioning_bucket.go:
```diff
@@ -2,11 +2,14 @@ package orm
 	"github.com/iov-one/weave/errors"
 )
 
+const versionMaxBytes = 4
```
I think this is not max bytes, but exact size in bytes of the serialized version number.

ruseinov replied:
Indeed, iterating on that constant's name made it worse.

ruseinov commented:
I have addressed some of the comments and replied to others. 

I think this might be ready for the final checkup.

@alpe I'd really like you to look at 4 byte vs 8 byte issue for versions, because if we go with 4 byte - we won't be able to go 8 byte without converting the whole database anymore.

ruseinov commented:
There is some weird issue with CI that does not have anything to do with my changes.
It also works locally, curious what causes that.

ruseinov commented:
Addressed!
